man/ipinfo.1
.TH ipinfo 1 "November 2007" minitools
.
.SH NAME
ipinfo \- Show IPv4 address information
.
.SH SYNOPSIS
.nf
\fBipinfo\fP [-hV] \fIaddr\fP/\fIn\fP
\fBipinfo\fP [-hV] \fIaddr\fP [\fImask\fP]
.fi
.
.SH DESCRIPTION
Given an IPv4 address \fIaddr\fP and, optionally,
a "slash" value /\fIn\fP or a (net or host) \fImask\fP,
compute and show information about this address:
netmask, hostmask, broadcast address, network address,
and maximum number of hosts in this network, both in
dotted decimal and in binary notation.
Also indicate the address class (if neither /\fIn\fP
nor a \fImask\fP is specified) and whether or not it
is a private address according to RFC 1918.
.
.SH OPTIONS
.TP 5
.B -h
Show quick help to standard output and quit.
.TP 5
.B -V
Show version to standard output and quit.
.
.SH EXAMPLE
.RB "$ " "ipinfo 192.168.25.108/27"
.nf
CIDR 27 host address, private
Address:    192.168.25.108   11000000.10101000.00011001.011/01100
Netmask:    255.255.255.224  11111111.11111111.11111111.111/00000  27
Hostmask:   0.0.0.31         00000000.00000000.00000000.000/11111   5
MaxHosts:   30               --------+--------+--------+---/-----
Network:    192.168.25.96    11000000.10101000.00011001.011/00000 min
Broadcast:  192.168.25.127   11000000.10101000.00011001.011/11111 max
.fi
.
.SH REMARKS
Originally, the first few bits of an IPv4 address decided about
its class, which defined how many bits designate the network
and how many bits designate the host within that network.
RFC 1519 on Classless Inter-Domain Routing (CIDR)
proposed a more flexible way of separating an IPv4 address
into a network part and a host part. With CIDR, a netmask (or,
equivalently, a hostmask) is used to make routing decisions.
.PP
An IPv4 \fBnetmask\fP specifies those bits of an IPv4 address
that designate the network part of the address. The network part
always consists of the \fIn\fP leftmost bits of the address.
For example, 255.255.255.224 designates the 27 leftmost bits.
.PP
A \fBhostmask\fP (or "wildcard" in Cisco terms) is the
bitwise negation of the netmask; it specifies those bits of
an IPv4 address that designate the host part of the address.
The host part always consists of the 32-\fIn\fP rightmost
bits of the address. For example, 0.0.0.31 designates the
5 rightmost bits.
.PP
Each valid netmask can also be represented more concisely
by the number \fIn\fP of bits that form the network part
of the address.  This number is known as the "slash-value"
because it is usually appended to an IPv4 address, separated
from it by a slash. For example, 192.168.3.224/27 means
that the first (leftmost) 27 bits are the network part
and the remaining (rightmost) 32-27=5 bits are the
host number within that network.
.PP
The first address within a subnet designates this network
itself. The last address within a subnet is the broadcast
address, used for sending packets to all hosts in the subnet.
Therefore, the number of addresses available for actual hosts
in a subnet is two less than 2 to the power of the number of
host bits. The first usable address is one more than the
network address, and the last usable address is one less
than the broadcast address.
.PP
Common CIDR suffixes and subnet masks, along with
the maximum number of hosts in such a network:
.PP
.nf
CIDR  Netmask         Hosts
/22   255.255.252.0    1022
/23   255.255.254.0     510
/24   255.255.255.0     254
/25   255.255.255.128   126
/26   255.255.255.192    62
/27   255.255.255.224    30
/28   255.255.255.240    14
/29   255.255.255.248     6
/30   255.255.255.252     2
.fi
.
.SH AUTHOR
Written by UJR in 2001, last updated in 2007.
.br
Released under the GNU General Public License (GPL).

.SH SEE ALSO
.nf
RFC 1918 Address Allocation for Private Internets
RFC 1519 Classless Inter-Domain Routing (CIDR)
.fi


src/ipinfo.c
/* Show IPv4 address information.
 * Usage: ipinfo [-hV] address[/n] [mask]
 * History:
 *   ujr/2001-10-25 started
 *   ujr/2002-09-03 added command line switches for selective output
 *   ujr/2005-04-10 changed -v to -V and my related standard behaviour
 *   ujr/2007-11-30 major rewrite, renamed ipcalc to ipinfo.
 * Note: works internally with a hostmask (not netmask).
 * License: GNU General Public License (GPL).
 */

#include <limits.h>
#include <stdio.h>
#include <string.h>

#include "common.h"

static char id[] = "This is ipinfo, version 1.0\n"
                   "Copyright (c) 2001-2007 by UJR\n";

static int identity(void);
static int usage(const char *errmsg);

static char getclass(uint32 ip);
static uint32 getmask(char class);
static int ispriv(uint32 ip);
static int mask2bits(uint32 mask);

static int scanip4(const char *s, uint32 *ip);
static const char *fmtip2(uint32 ip, int slash);
static const char *fmtsep(int slash);

static char *me = "ipinfo";

int main(int argc, char *argv[])
{
  const char *addrstr;
  const char *maskstr;
  uint32 addr, mask;
  int c, i, j, slash = -1; /* no slash */
  char ipclass = 0; /* classless */
  uint32 nwaddr, bcaddr, count;

  (void) argc; /* unused */
  if (argv && *argv) me = *argv;
  else return FAILHARD; /* no arg0? */

  while (*++argv && (**argv == '-')) {
    while ((c = *++argv[0])) switch (c) {
      case 'h': return usage(0);
      case 'V': return identity();
      case '-': argv++; goto endargs;
      default: return usage("invalid option");
    }
  }
endargs:

  if (*argv) addrstr = *argv++;
  else return usage("no address specified");

  if (*argv) maskstr = *argv++;
  else maskstr = 0; /* not specified */

  if (*argv) return usage("too many arguments");

  /* Parse address and opt slash value */

  i = scanip4(addrstr, &addr);
  if (i && addrstr[i] == '/') {
    i += 1; /* skip slash */
    j = scanuint(addrstr+i, (unsigned*)&slash);
    if (j) i += j; else i = 0;
  }
  else slash = -1; /* not specified */

  if (i == 0 || addrstr[i] != '\0')
    return usage("invalid address");
  if (slash > 32)
    return usage("slash value out of range 0..32");

  if (slash < 0) { /* negative means classes */
    ipclass = getclass(addr);
    mask = getmask(ipclass);
  }
  else {
    ipclass = 0; /* classless address (CIDR) */
    mask = (1 << (32-slash))-1;
  }

  /* Parse optional net/host mask or assume class */

  if (maskstr) {
    if (slash >= 0)
      return usage("too many arguments");
    i = scanip4(maskstr, &mask);
    if ((i == 0) || (maskstr[i] != '\0'))
      return usage("invalid mask");
    ipclass = 0; /* explicit mask means no class (CIDR) */

    /* If the given mask appears to be a netmask, convert
     * to hostmask (Cisco wildcard) by bitwise negation.
     */
    if (mask & 0x80000000) mask = ~mask;
  }

  /* Compute the "slash-value" corresponding to mask;
   * check if valid mask (ie, mask+1 is power of two).
   */
  if (slash < 0) /* compute netbits from mask */
    if ((slash = 32 - mask2bits(mask)) > 32)
      return usage("invalid mask");

  /* Compute other interesting values and output */

  bcaddr = addr | mask;
  mask = ~mask; /* convert to netmask */
  nwaddr = addr & mask;
  count = (1 << (32 - slash)) - 2;

  if (ipclass) printf("Class %c", ipclass);
  else printf("CIDR %d", slash);
  printf(" %s address", (nwaddr == addr) ? "network"
      : (bcaddr == addr) ? "broadcast" : "host");
  if (ispriv(addr)) printf(", private");
  printf("\n");

  printf("Address:    %s\n", fmtip2(addr, slash));
  printf("Netmask:    %s %3d\n", fmtip2(mask, slash), slash);
  printf("Hostmask:   %s %3d\n", fmtip2(~mask, slash), 32-slash);
  printf("MaxHosts:   %-17ld%s\n", (long) count, fmtsep(slash));
  printf("Network:    %s min\n", fmtip2(nwaddr, slash));
  printf("Broadcast:  %s max\n\n", fmtip2(bcaddr, slash));

  return SUCCESS;
}

static int identity(void)
{
  return fputs(id, stdout) >= 0 ? SUCCESS : FAILSOFT;
}

static int usage(const char *errmsg)
{
  FILE *fp = errmsg ? stderr : stdout;
  if (errmsg) fprintf(fp, "%s: %s\n", me, errmsg);
  else fprintf(fp, "Show IPv4 address information\n");
  fprintf(fp, "Usage: %s [-V] address/n\n", me);
  fprintf(fp, "   or: %s [-V] address [mask]\n", me);
  return errmsg ? FAILHARD : SUCCESS;
}

/** Return the class for an IPv4 address */
static char getclass(uint32 ip)
{
  int first = (ip >> 24) & 255; /* first octet */
  if ((first & 0x80) == 0x00) return 'A';
  if ((first & 0xc0) == 0x80) return 'B';
  if ((first & 0xe0) == 0xc0) return 'C';
  if ((first & 0xf0) == 0xe0) return 'D';
  return 'E'; /* class E, reserved */
}

/** Return the hostmask for a class */
static uint32 getmask(char class)
{
  if (class == 'A') return 0x00ffffff;
  if (class == 'B') return 0x0000ffff;
  if (class == 'C') return 0x000000ff;
  return 0; /* no hostmask for other classes */
}

/** Return true iff the given IP address is a private address,
 * ie, in one of the networks 10/8 or 172.16/12 or 192.168/16.
 */
static int ispriv(uint32 ip)
{
  int x = ip >> 16; /* first two octets decide */
  return ((x & 0xff00) == (uint32) 10*256) ||
         ((x & 0xfff0) == (uint32) 172*256+16) ||
         ((x & 0xffff) == (uint32) 192*256+168);
}

/** Count trailing one bits (mask to bit count).
 * By the way, the inverse is mask=(1<<bits)-1. */
static int mask2bits(uint32 mask)
{
  int n;

  if (mask == 0) return 32; /* 2^32 = 0 (mod wordsize) */
  if (mask & (mask+1)) return -1; /* mask is not 2^n-1 */

  /* count 1-bits from the right */
  for (n=0; mask & 1; n++) mask >>= 1;

  return n;
}

/** Scan an IPv4 addr in dotted decimal notation, return #chars scanned */
static int scanip4(const char *s, uint32 *ip)
{
  const char *p = s;
  uint32 value;
  unsigned octet;
  int n;

  if ((n = scanuint(p, &octet)) == 0 || (octet > 255)) return 0;
  value = (uint32) octet;  p += n;

  if (*p++ != '.') return 0;

  if ((n = scanuint(p, &octet)) == 0 || (octet > 255)) return 0;
  value = 256*value + octet;  p += n;

  if (*p++ != '.') return 0;

  if ((n = scanuint(p, &octet)) == 0 || (octet > 255)) return 0;
  value = 256*value + octet;  p += n;

  if (*p++ != '.') return 0;

  if ((n = scanuint(p, &octet)) == 0 || (octet > 255)) return 0;
  value = 256*value + octet;  p += n;

  if (ip) *ip = value;
  return p - s; /* #chars scanned */
}

static const char *fmtip2(uint32 ip, int slash)
{
  int i, j;
  static char buf[80];

  i = sprintf(buf, "%d.%d.%d.%d", /* dotted decimal */
    (int) (ip>>24)&255, (int) (ip>>16)&255, (int) (ip>>8)&255, (int) ip&255);
  for (; i < 17; i++) buf[i] = ' ';
  slash = 32 - slash;
  for (j=31; j>=0; j--) { /* binary representation */
    buf[i++] = (ip & (1<<j)) ? '1' : '0';
    if (j == slash) buf[i++] = '/';
    else if (j == 24 || j == 16 || j == 8) buf[i++] = '.';
  }
  buf[i] = '\0';

  return (const char *) buf;
}

static const char *fmtsep(int slash)
{
  int i, j = 0;
  static char buf[40];

  for (i = 1; i <= 32; i++) {
    buf[j++] = '-';
    if (i == slash) buf[j++] = '/';
    else if (i == 24 || i == 16 || i == 8) buf[j++] = '+';
  }
  buf[j] = '\0';

  return (const char *) buf;
}
